src/main/java/org/jenkinsci/plugins/workflow/flow/StashManager.java

```diff
@@ -228,16 +286,77 @@ public boolean shouldClearAll(@Nonnull Run<?,?> build) {
 
     }
 
+    /**
+     * Mixin interface for an {@link ArtifactManager} which supports specialized stash behavior as well.
+     */
+    public interface StashAwareArtifactManager /* extends ArtifactManager */ {
+
+        /** @see StashManager#stash(Run, String, FilePath, Launcher, EnvVars, TaskListener, String, String, boolean, boolean) */
+        void stash(@Nonnull String name, @Nonnull FilePath workspace, @Nonnull Launcher launcher, @Nonnull EnvVars env, @Nonnull TaskListener listener, @CheckForNull String includes, @CheckForNull String excludes, boolean useDefaultExcludes, boolean allowEmpty) throws IOException, InterruptedException;
+
+        /** @see StashManager#unstash(Run, String, FilePath, Launcher, EnvVars, TaskListener) */
+        void unstash(@Nonnull String name, @Nonnull FilePath workspace, @Nonnull Launcher launcher, @Nonnull EnvVars env, @Nonnull TaskListener listener) throws IOException, InterruptedException;
+
+        /** @see StashManager#clearAll(Run, TaskListener) */
+        void clearAllStashes(@Nonnull TaskListener listener) throws IOException, InterruptedException;
+
+        /**
+         * Copy all stashes and artifacts from one build to another.
+         * The {@link ArtifactManager} configuration will be as of the origin build.
+         * If the implementation cannot handle {@code to} for whatever reason, it may throw {@link AbortException}.
+         * @see CopyStashesAndArtifacts
+         */
+        void copyAllArtifactsAndStashes(@Nonnull Run<?,?> to, @Nonnull TaskListener listener) throws IOException, InterruptedException;
+
+    }
+
+    private static @CheckForNull StashAwareArtifactManager stashAwareArtifactManager(@Nonnull Run<?, ?> build) throws IOException {
+        ArtifactManager am = build.pickArtifactManager();
+        return am instanceof StashAwareArtifactManager ? (StashAwareArtifactManager) am : null;
+    }
+
+    private static @Nonnull Launcher launcherFor(@Nonnull FilePath workspace, @Nonnull TaskListener listener) {
+        Computer c = workspace.toComputer();
+        if (c != null) {
+            Node n = c.getNode();
+            if (n != null) {
+                return n.createLauncher(listener);
+            } else {
+                listener.error(c.getDisplayName() + " seems to be offline");
+                return new LocalLauncher(listener);
+            }
+        } else {
+            listener.error(workspace + " seems to be offline");
+            return new LocalLauncher(listener);
+        }
+    }
+
+    private static @Nonnull EnvVars envFor(@Nonnull FilePath workspace, @Nonnull TaskListener listener) throws IOException, InterruptedException {
```

Comment:
Not sure about the current implementation. It builds computer environment variables, but ignores possible overrides by the `Run`'s logic. I hope it does not impact artifact managers, but I can imagine `withEnvVars()` closure overriding some variables for the external artifact manager. E.g. just `tool()` installation for the artifact management binary.

Maybe an edge case

Reply:
Yes this could be improved I think—will check. Anyway this method is only called when using the deprecated overloads. As of https://github.com/jenkinsci/workflow-basic-steps-plugin/pull/60 the actual `EnvVars` is loaded from the step context, for example taking into account `withEnv` blocks.